### PR TITLE
ButtonPullDown class option

### PR DESCRIPTION
**Pull request by artandtechvof**
My second attempt at a pull request for Button pull-down option. This time made a separate class of it and tested for single dual and quad button setups. They are all functioning as expected, and compiler indicates the following additional overhead generated by this class (it even saves some bytes for 4 button setups...);

single button
-no difference between both classes, Sketch uses 201256 bytes (81%) of program storage space. Maximum is 245760 bytes.

2 button setup
-original: Sketch uses 201368 bytes (81%) of program storage space. Maximum is 245760 bytes.
-2x ButtonPullDown Sketch uses 201368 bytes (81%) of program storage space. Maximum is 245760 bytes. -> exact same
-1x Button, 1x ButtonPullDown Sketch uses 201480 bytes (81%) of program storage space. Maximum is 245760 bytes. -> 112 bytes extra

4button setup 
- 4 buttons original class,  Sketch uses 201456 bytes (81%) of program storage space. Maximum is 245760 bytes. total bytes unchanged
- 4 buttons all ButtonPullDown type,   Sketch uses 201448 bytes (81%) of program storage space. Maximum is 245760 bytes.  -> 8 bytes less (!)
- only 1 button uses ButtonPullDown class,  Sketch uses 201632 bytes (82%) of program storage space. Maximum is 245760 bytes.  -> 176 extra
- 2x Buttons, 2x ButtonPullDown,  Sketch uses 201680 bytes (82%) of program storage space. Maximum is 245760 bytes. -> 224 bytes extra
- 3 buttons use ButtonPullDown type,  Sketch uses 201632 bytes (82%) of program storage space. Maximum is 245760 bytes. -> 176 bytes extra

### CODE REVIEW

File: buttons/button.h [hunk 1]
```diff
@@ -3,19 +3,65 @@
...
   bool Read() override {
     return digitalRead(pin_) == LOW;
   }
+};
+
+
+// For special setups were you want to hook up a momentary button to VCC(3.3V) or Vbat
+// and the other end to the inputs of the Proffieboard / Teensy 
+// 
+// For Proffieboards all inputs can handle Vbat, except for Data3 / blade3pin
```
Comment (profezzorn) [inline buttons/button.h:36]: This is actually different for Proffieboard1 and Proffieboard2
Also, I thought it was Data4?

File: buttons/button.h [hunk 2]
```diff
@@ -3,19 +3,65 @@
...
+//
+// This class will add some additional bytes to your code,
```
Comment (profezzorn) [inline buttons/button.h:38]: The size comments are goign to depend on too many variables to be accurate.
The fact that using more things takes more memory is pretty obvious, so I don't think we need to say that either.

File: buttons/button.h [hunk 3]
```diff
@@ -3,19 +3,65 @@
...
+    pinMode(pin, INPUT_PULLDOWN);
+#ifdef ENABLE_SNOOZE
+    snooze_digital.pinMode(pin, INPUT_PULLDOWN, FALLING);
+#endif
+  }
+protected:
   uint8_t pin_;
+
+  bool Read() override {
+	 return digitalRead(pin_) == HIGH;
```
Comment (profezzorn) [inline buttons/button.h:60]: weird indentation

File: buttons/button.h [hunk 4]
```diff
@@ -3,19 +3,65 @@
...
+  }
+
```
Comment (profezzorn) [inline buttons/button.h:62]: extra empty line

File: buttons/button.h [hunk 5]
```diff
@@ -3,19 +3,65 @@
...
   uint8_t pin_;
+
+  bool Read() override {
+	 return digitalRead(pin_) == HIGH;
+  }
+   
 };
 
 #endif
+
```
Comment (profezzorn) [inline buttons/button.h:66]: two extra empty lines

File: buttons/button.h [hunk 6]
```diff
@@ -3,19 +3,65 @@
...
+// (224 bytes, for a 4 button setup with 2 button pullup, and 2 buttons pulldown)
+// However, if all of your used switches are made ButtonPullDown 
+// it will save you a couple of bytes compared to the original Button class 
+// (8 bytes for 4 button setup)
+//
+// Usage example, put this in the button config section: 
+// ButtonPullDown PowerButton(BUTTON_POWER, powerButtonPin, "pow");  
+//
+
+class ButtonPullDown : public ButtonBase{
```
Comment (profezzorn) [inline buttons/button.h:48]: ButtonPullDown sounds weird to me. Like it's a pull-down for a button, not a button that has a pull-down.
Examples of other class names in this directory: LatchingButton, FloatingButton, DebouncedButton, TouchButton
So, for consistency, I think this class should be called PullDownButton.

File: buttons/button.h (hunk 1 above)
Comment (artandtechvof) [reply buttons/button.h:36]: OK, should non 5V tolerant pins be mentioned for both version? Or just leave this out of the comments?
Proffieboard V2, it is actualy Data3 / blade3pin according to DS11449 Rev 5 STM32L433xx
 
/ Proffieboard pin map  , 5V toleracny according to DS11449 Rev 5 STM32L433xx FT = tolerant, TT is 3.6 tolerant
enum SaberPins {
  // I2S
  bclkPin = 3,                   // BCLK (digital audio)   PB13 FT_fl
  txd0Pin = 26,                  // TXD0 (digital audio)   PA10 FT_fl
  lrclkPin = 2,                  // LRCLK (digital audio)  PB12 FT_I

  // I2C
  i2cDataPin = 7,                 // I2C bus, Used by motion sensors  PB9 FT_fl
  i2cClockPin = 25,               // I2C bus, Used by motion sensors  PA9 FT_fl

  // Buttons
  powerButtonPin = 21,            // power button  PB6 FT_I
  auxPin = 23,                    // AUX button    PB5 FT_l
  aux2Pin = 22,                   // AUX2 button   PB4 FT_fla

  // Memory card
  sdCardSelectPin = 4,            // PB14 FT_fl

  amplifierPin = 24,              // Amplifier enable pin PH1 FT
  boosterPin = 15,                // Booster enable pin   PH0 FT
  motionSensorInterruptPin = 12,  // motion sensor interrupt PC13 FT

  // No fastled support yet
  spiLedSelect = -1,
  spiLedDataOut =-1,
  spiLedClock = -1,

  // Neopixel pins
  bladePin = 16,                  // blade control, either WS2811 or PWM PA0  FT_a
  bladeIdentifyPin = 16,          // blade identify input / FoC  
  blade2Pin = 1,                  // PB10   FT_fl
  blade3Pin = 17,                 // PA4  -> not 5V tolerant  TT_a
  blade4Pin = 0,                  // PB3  FT_la	
  blade5Pin = 8,                  // PA15 (also UART)  FT_li
  blade6Pin = 9,                  // PA02 (also UART)  FT_la

  // Blade power control
  bladePowerPin1 = 13,            // blade power control PA1  FT_la
  bladePowerPin2 = 19,            // blade power control PB8  FT_fl
  bladePowerPin3 = 18,            // blade power control PA8  FT_l
  bladePowerPin4 = 10,            // blade power control PB3  FT_la
  bladePowerPin5 = 5,             // blade power control PB15 FT_I
  bladePowerPin6 = 6,             // blade power control PB0  FT_la

  // If there is no neopixels, these can be used as PWM output
  // If neopixels are present, but not on these pins, PWM may still
  // be possible at 800kHz.
  // You can drive 20mA LEDs with a resistor from these pins, or
  // hook up an external FET to drive more powerful LEDs
  bladePowerPin7 = 17,            // PB03 (also neopix4)
  bladePowerPin8 = 1,             // PB10 (also neopix2)
  bladePowerPin9 = 0,             // PB11 (also neopix3)
  bladePowerPin10 = 16,           // PA00 (also blade ID / neopixels)

  // Analog pins
  batteryLevelPin = 14,           // battery level input PA04 TT_a  -> Not 5V tolerant, not broken out to any pads on proffieboard

  // UART
  rxPin = 8,                      // PA15  FT_I
  txPin = 9,                      // PA02  FT_la
};

File: buttons/button.h (hunk 2 above)
Comment (artandtechvof) [reply buttons/button.h:38]: Understood, I will delete these notes.

File: buttons/button.h (hunk 3 above)
Comment (artandtechvof) [reply buttons/button.h:60]: I will correct this

File: buttons/button.h (hunk 4 above)
Comment (artandtechvof) [reply buttons/button.h:62]: I will correct this

File: buttons/button.h (hunk 5 above)
Comment (artandtechvof) [reply buttons/button.h:66]: I will correct this

File: buttons/button.h (hunk 6 above)
Comment (artandtechvof) [reply buttons/button.h:48]: Understood. I will rename the function to PullDownButton

**Comment by artandtechvof** [comment]
Thank you for reviewing the code and helping to make it suitable for merging! I am a little rusty in using github, commits and pull-requests as you might have noticed... Learning fast though.

File: buttons/button.h [hunk 7]
```diff
@@ -3,19 +3,56 @@
...
     return digitalRead(pin_) == LOW;
   }
-  uint8_t pin_;
 };
 
+
+// For special setups were you want to hook up a momentary button to VCC(3.3V) or Vbat
+// and the other end to the inputs of the Proffieboard / Teensy 
+// 
+// For Proffieboards V2 all inputs can handle Vbat, except for Data3 / blade3pin
```
Comment (profezzorn) [inline buttons/button.h:36]: This is incorrect, Data4 is the one that not 5V tolerant.

File: buttons/button.h [hunk 8]
```diff
@@ -3,19 +3,56 @@
...
+  }
+protected:
+  uint8_t pin_;
+
+  bool Read() override {
+	return digitalRead(pin_) == HIGH;
+  }
+};
+
+#endif
```
Comment (profezzorn) [inline buttons/button.h:58]: Missing newline at end of file.

Comment (artandtechvof) [reply buttons/button.h]: Newline added

File: buttons/button.h (hunk 7 above)
Comment (artandtechvof) [reply buttons/button.h:36]: As stated in the V2 config file;
(or did you reroute the pins different?)

blade3Pin = 17, // PA4 -> not 5V tolerant TT_a
blade4Pin = 0, // PB3 FT_la

From the DS11449 Rev 5 STM32l433cc data sheet;

14 14 G6 E5 20 20 H3 29 M3 **_PA4_** I/O **_TT_a_** -SPI1_NSS,SPI3_NSS,USART2_CK,SAI1_FS_B,LPTIM2_OUT,EVENTOUT 
COMP1_INM,
COMP2_INM,
ADC1_IN9,
DAC1_OUT1

39 39 A3 A5 55 54 A5 89 A8 **_PB3_** (JTDOTRACESWO) I/O **_FT_la_** (3) JTDO-TRACESWO,TIM2_CH2,SPI1_SCK,SPI3_SCK, USART1_RTS_DE,
LCD_SEG7,
SAI1_SCK_B,
EVENTOUT
COMP2_INM

Comment (profezzorn) [reply buttons/button.h:36]: The comments are not accurate it seems.  (I will fix that)
blade4Pin is PA4 and blade3Pin is PB3

Comment (artandtechvof) [reply buttons/button.h]: Ah, That makes sense. Also the fact that Data4 is the only DAC capable output and PA4 being only 1 of 2 pins that can do this...

Modified the comment to Data4 not being 5V tolerant.

**Pull request merged**